File: test/uint256_t.go
package main

import (
	"fmt"
	"github.com/holiman/uint256"
)

func load_u256(u *uint256.Int, i int) {
	u.Set(test_cases_u256[i])
}

func print_uint(label string, x *uint256.Int) {
	fmt.Println(label+"[0]:", x[0])
	fmt.Println(label+"[1]:", x[1])
	fmt.Println(label+"[2]:", x[2])
	fmt.Println(label+"[3]:", x[3])
}

func tf(v bool) int {
	if v {
		return 1
	}
	return 0
}

func opSHL(res, x, y *uint256.Int) {
	if y.LtUint64(256) {
		res.Lsh(x, uint(y.Uint64()))
	} else {
		res.Clear()
	}
}

func opSHR(res, x, y *uint256.Int) {
	if y.LtUint64(256) {
		res.Rsh(x, uint(y.Uint64()))
	} else {
		res.Clear()
	}
}

func opSAR(res, value, shift *uint256.Int) {
	if shift.GtUint64(256) {
		if value.Sign() >= 0 {
			res.Clear()
		} else {
			res.SetAllOne()
		}
		return
	}
	n := uint(shift.Uint64())
	res.SRsh(value, n)
}

func test_run(i int, res, x, y, z *uint256.Int) {
    fmt.Println("----------------------------------------", i);

    // add
    res.Add(x, y);
    print_uint("add(x, y)", res);

    // mul
    res.Mul(x, y);
    print_uint("mul(x, y)", res);

    // sub
    res.Sub(x, y);
    print_uint("sub(x, y)", res);

    // div
    res.Div(x, y);
    print_uint("div(x, y)", res);

    // sdiv
    res.SDiv(x, y);
    print_uint("sdiv(x, y)", res);

    // mod
    res.Mod(x, y);
    print_uint("mod(x, y)", res);

    // smod
    res.SMod(x, y);
    print_uint("smod(x, y)", res);

    // add_mod
    res.AddMod(x, y, z);
    print_uint("add_mod(x, y, z)", res);

    // mul_mod
    res.MulMod(x, y, z);
    print_uint("mul_mod(x, y, z)", res);

    // _exp
    res.Exp(x, y);
    print_uint("_exp(x, y)", res);

    // sign_extend
    res.ExtendSign(x, y);
    print_uint("sign_extend(x, y)", res);

    // lt
    r := x.Lt(y);
    fmt.Println("lt(x, y):", tf(r));

    // gt
    r = x.Gt(y);
    fmt.Println("gt(x, y):", tf(r));

    // slt
    r = x.Slt(y);
    fmt.Println("slt(x, y):", tf(r));

    // sgt
    r = x.Sgt(y);
    fmt.Println("sgt(x, y):", tf(r));

    // eq
    r = x.Eq(y);
    fmt.Println("eq(x, y):", tf(r));

    // is_zero
    r = res.IsZero();
    fmt.Println("is_zero(x):", tf(r));

    // and
    res.And(x, y);
    print_uint("and(x, y)", res);

    // or
    res.Or(x, y);
    print_uint("or(x, y)", res);

    // xor
    res.Xor(x, y);
    print_uint("xor(x, y)", res);

    // not
    res.Not(x);
    print_uint("not(x)", res);

    byte_check := &uint256.Int{x[0], x[1], x[2], x[3]}
    // byte
    byte_check.Byte(y);
    print_uint("byte(x, y)", byte_check);

    // shl
    opSHL(res, x, y);
    print_uint("shl(x, y)", res);

    // shr
    opSHR(res, x, y);
    print_uint("shr(x, y)", res);

    // sar
    opSAR(res, x, y);
    print_uint("sar(x, y)", res);
}

func main() {
	var res *uint256.Int = &uint256.Int{}
	var x *uint256.Int = &uint256.Int{}
	var y *uint256.Int = &uint256.Int{}
	var z *uint256.Int = &uint256.Int{}

	for i:=0; i < len(test_cases_u256) - 3; i+= 3 {
		load_u256(x, i)
		load_u256(y, i+1)
		load_u256(z, i+2)
		test_run(i/3, res, x, y, z)
	}

    // sm := uint256.NewInt(1)
    // bi := uint256.NewInt(2)
    // sub := sm.Sub(sm, bi)
    // fmt.Println(sub.Dec())
}

File: test/uint256.t.js
const fs = require("fs");
const ethers = require("ethers");
const crypto = require("crypto");


const NUM_TESTS = 60;

const endpoint = process.env.ENDPOINT;

function randomBigInt() {
    let b = crypto.randomBytes(32);
    return BigInt('0x' + b.toString('hex'));
}

const opcodes = [
    'Add', 'Mul', 'Sub', 'Div', 'SDiv', 'Mod', 'SMod', 'AddMod', 'MulMod',
    'Exp', 'SignExtend', 'Lt', 'Gt', 'Slt', 'Sgt', 'Eq', 'IsZero', 'And', 'Or',
    'Xor', 'Not', 'Byte', 'Shl', 'Shr', 'Sar'
];

const bitmasks = [];
for (let i = 0; i < 25; i++) {
    bitmasks[i] = 1n << BigInt(i);
}

// commenting out any of these values removes them from the test call
const opcodes_to_test = [
    'Add',
    'Mul',
    'Sub',
    'Div',
    'SDiv',
    'Mod',
    'SMod',
    'AddMod',
    'MulMod',
    'Exp',
    'SignExtend',
    'Lt',
    'Gt',
    'Slt',
    'Sgt',
    'Eq',
    'IsZero',
    'And',
    'Or',
    'Xor',
    'Not',
    'Byte',
    'Shl',
    'Shr',
    'Sar'
];

var tests_map = 0n;
for (let i = 0; i < opcodes_to_test.length; i++) {
    const bitmask_index = opcodes.indexOf(opcodes_to_test[i]);
    tests_map |= BigInt(bitmasks[bitmask_index]);
}

async function main() {
    const provider = new ethers.JsonRpcProvider(endpoint)

    const data = JSON.parse(fs.readFileSync('./test/test_address.json'));
    const addr = data.address;
    console.log("address: ", addr);

    const contract_data = JSON.parse(fs.readFileSync('build/interface.json'));
    const abi = contract_data.contracts.uint256_test.Uint256Test.abi;
    const contract = new ethers.Contract(addr, abi, provider);

    const values = [];
    for (let i = 0; i<NUM_TESTS*3; i++) {
        values[i] = randomBigInt();
    }
    // the tests work by using the three inputs:
    // x:values[0], y:values[1], z:values[2]
    // x:values[3], y:values[4], z:values[5]
    // and so on...
    // the expected result is determined by the appropriate EVM opcode, e.g.
    // want := addmod(x, y, z) or want := add(x, y)
    // z is ignore for tests that only take two arguments.

    // restrict some of the y values to be less than 255 and 30, because some
    // tests are basically meaningless outside of this range. in particular,
    // shl, shr, sar, byte

    const first_sixth = NUM_TESTS/6;
    // first 1/6 of tests are 255 limit
    for (let i = 0; i < first_sixth; i++) {
        values[i*3 + 1] %= 255n;
    }
    // second 1/6 of tests are 30 limit
    for (let i = first_sixth; i < first_sixth*2; i++) {
        values[i*3 + 1] %= 30n;
    }

    const test_result = await contract.test_values(values, tests_map);

    const [passing, opcode, have, want, x, y, z] = test_result;
    if (passing) {
        console.log('Test passed!');
    } else {
        console.log("Test failed. Here's the data:");
        console.log(`Opcode: ${opcodes[opcode]}`);
        console.log(`have: ${have}`);
        console.log(`want: ${want}`);
        console.log(`x: ${x}`);
        console.log(`y: ${y}`);
        console.log(`z: ${z}`);
    }
}

main()
